### PR TITLE
Fix dead bodies pulling and dragging

### LINKED ISSUE #3570: Unable to drag or click-and-drag on dead bodies
I had initially reported this under #3464 but realized that it actually has nothing to do with gravity being on or off. Bodies can no longer be dragged at all, and can't be click-and-dragged into the body scanner even if you bring the scanner to the body. In some cases they also cannot be looted via click-and-drag.

This occurs even if gravity is never turned off and someone happens to die at the start of a new round.

### PR DESCRIPTION
**Pull request by Macoron**
## About the PR 
Fixes #3570 

**Screenshots**

**Changelog**

:cl:
- fix: Fix dead bodies pulling and dragging. No more chair ambulance.

### CODE REVIEW

File: Content.Server/GameObjects/Components/Mobs/State/DeadMobState.cs [hunk 1]
```diff
@@ -33,7 +34,10 @@ public override void EnterState(IEntity entity)
 
             if (entity.TryGetComponent(out IPhysBody? physics))
             {
-                physics.CanCollide = false;
+                // remove mob from layers where they can be hit by ranged weapon
+                // other layers will stay unchanged
+                foreach (var fixture in physics.Fixtures)
+                    fixture.CollisionLayer &= (int) ~CollisionGroup.WeaponMask;
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Mobs/State/DeadMobState.cs:40]: Should get a specific fixture IMO.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
This doesn't fix the actual underlying issue for joints so the ticket shouldn't be closed. It'd also be better if we kept the behavior similar to items in terms of collision long-term but it's fine for now.

**Comment by Macoron** [comment]
> This doesn't fix the actual underlying issue for joints so the ticket shouldn't be closed.
Can you be more specific about what kind of issue are you talking about?

File: Content.Shared/Physics/CollisionGroup.cs [hunk 2]
```diff
@@ -28,6 +28,12 @@ public enum CollisionGroup
 
         MobMask = Impassable | MobImpassable | VaultImpassable | SmallImpassable,
         ThrownItem = MobImpassable | Impassable,
+
+        ProjectileMask = Impassable | MobImpassable, // will be damaged by bullets, arrows or other projectile
+        LaserMask = Opaque, // will be damaged by laser ray from laser gun
```
[the remark this answers is not in the document]
Comment (Macoron) [reply Content.Shared/Physics/CollisionGroup.cs:33]: Do you suggest use `WeaponMask = Opaque`?

File: Content.Server/GameObjects/Components/Mobs/State/DeadMobState.cs (hunk 1 above)
Comment (Macoron) [reply Content.Server/GameObjects/Components/Mobs/State/DeadMobState.cs:40]: By fixture name?

File: Content.Shared/Physics/CollisionGroup.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (metalgearsloth) [reply Content.Shared/Physics/CollisionGroup.cs:33]: WeaponMask = ProjectileMask | Opaque is probably fine

**Comment by metalgearsloth** [comment]
> 
> 
> > This doesn't fix the actual underlying issue for joints so the ticket shouldn't be closed.
> 
> Can you be more specific about what kind of issue are you talking about?

Joints currently isn't handled properly when collision is turned on / off which is why pulling crit bodies is okay but dead bodies isn't as it gets scuffed during the state change.

This was why I hadn't prd a fix earlier as it needs engine work.

File: Content.Server/GameObjects/Components/Mobs/State/DeadMobState.cs (hunk 1 above)
Comment (metalgearsloth) [reply Content.Server/GameObjects/Components/Mobs/State/DeadMobState.cs:40]: Yeah probably, just add a name to mobs for their default circle fixture and refer to it by that name.

**Comment by Macoron** [comment]
> Joints currently aren't handled properly when collision is turned on / off

I don't understand. When you disable collision off, body should go through walls (imagine pulling in Zero-G). So disabling collision for dead bodies or enable/disable it while pulling sounds like a really bad idea.

**Comment by 20kdc** [comment]
Did some testing on master (not with this PR):
I don't think there is an underlying issue with joints, at least at this time.
Right now, `CanCollide` is a misnomer - it really controls most physics stuff involving the object.
And players don't have `CollisionWakeComponent` to wake them back up.
So turning `CanCollide` to false disables their physics.
Now, adding CollisionWakeComponent, then disabling and then enabling it (I don't know why this is needed), "fixes" the dead body. Pulling it works, for example.
Of note is that (as expected) CanCollide is false when the body is not awake.

To me, all of this seems to be expected behaviour assuming `CanCollide` would be better named as `Enabled`, noted as the Box2D name for it.

**Comment by metalgearsloth** [comment]
You still working on this? The other PR got merged so just needs the fixture work or the likes to tweak it.

**Comment by Macoron** [comment]
Sadly, don't have time to finish it, so will close it

**Pull request closed without merging**